File: src/ui/clickable_elems.cpp
// Luanti
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright (C) 2024 v-rob, Vincent Robinson <robinsonvincent89@gmail.com>

#include "ui/clickable_elems.h"

#include "debug.h"
#include "log.h"
#include "ui/manager.h"
#include "util/serialize.h"

namespace ui
{
	void Button::reset()
	{
		Elem::reset();

		m_disabled = false;
	}

	void Button::read(std::istream &is)
	{
		auto super = newIs(readStr32(is));
		Elem::read(super);

		u32 set_mask = readU32(is);

		m_disabled = testShift(set_mask);

		if (testShift(set_mask))
			enableEvent(ON_PRESS);
	}

	bool Button::processInput(const SDL_Event &event)
	{
		return getMain().processFullPress(event, UI_CALLBACK(onPress));
	}

	void Button::onPress()
	{
		if (!m_disabled && testEvent(ON_PRESS)) {
			g_manager.sendMessage(createEvent(ON_PRESS).str());
		}
	}

	void Toggle::reset()
	{
		Elem::reset();

		m_disabled = false;
	}

	void Toggle::read(std::istream &is)
	{
		auto super = newIs(readStr32(is));
		Elem::read(super);

		u32 set_mask = readU32(is);

		m_disabled = testShift(set_mask);
		testShiftBool(set_mask, m_selected);

		if (testShift(set_mask))
			enableEvent(ON_PRESS);
		if (testShift(set_mask))
			enableEvent(ON_CHANGE);
	}

	bool Toggle::processInput(const SDL_Event &event)
	{
		return getMain().processFullPress(event, UI_CALLBACK(onPress));
	}

	void Toggle::onPress()
	{
		if (m_disabled) {
			return;
		}

		m_selected = !m_selected;

		// Send both a press and a change event since both occurred.
		if (testEvent(ON_PRESS)) {
			g_manager.sendMessage(createEvent(ON_PRESS).str());
		}
		if (testEvent(ON_CHANGE)) {
			auto os = createEvent(ON_CHANGE);
			writeU8(os, m_selected);

			g_manager.sendMessage(os.str());
		}
	}

	void Option::reset()
	{
		Elem::reset();

		m_disabled = false;
		m_family.clear();
	}

	void Option::read(std::istream &is)
	{
		auto super = newIs(readStr32(is));
		Elem::read(super);

		u32 set_mask = readU32(is);

		m_disabled = testShift(set_mask);
		testShiftBool(set_mask, m_selected);

		if (testShift(set_mask))
			m_family = readNullStr(is);

		if (testShift(set_mask))
			enableEvent(ON_PRESS);
		if (testShift(set_mask))
			enableEvent(ON_CHANGE);
	}

	bool Option::processInput(const SDL_Event &event)
	{
		return getMain().processFullPress(event, UI_CALLBACK(onPress));
	}

	void Option::onPress()
	{
		if (m_disabled) {
			return;
		}

		// Send a press event for this pressed option button.
		if (testEvent(ON_PRESS)) {
			g_manager.sendMessage(createEvent(ON_PRESS).str());
		}

		// Select this option button unconditionally before deselecting the
		// others in the family.
		onChange(true);

		// If this option button has no family, then don't do anything else
		// since there may be other buttons with the same empty family string.
		if (m_family.empty()) {
			return;
		}

		// If we find any other option buttons in this family, deselect them.
		for (Elem *elem : getWindow().getElems()) {
			if (elem->getType() != getType()) {
				continue;
			}

			Option *option = (Option *)elem;
			if (option->m_family == m_family && option != this) {
				option->onChange(false);
			}
		}
	}

	void Option::onChange(bool selected)
	{
		bool was_selected = m_selected;
		m_selected = selected;

		// If the state of the option button changed, send a change event.
		if (was_selected != m_selected && testEvent(ON_CHANGE)) {
			auto os = createEvent(ON_CHANGE);
			writeU8(os, m_selected);

			g_manager.sendMessage(os.str());
		}
	}
}

File: builtin/ui/clickable_elems.lua
-- Luanti
-- SPDX-License-Identifier: LGPL-2.1-or-later
-- Copyright (C) 2024 v-rob, Vincent Robinson <robinsonvincent89@gmail.com>

ui.Button = ui._new_type(ui.Elem, "button", 0x02, true)

function ui.Button:_init(props)
	ui.Elem._init(self, props)

	self._disabled = props.disabled
	self._on_press = props.on_press
end

function ui.Button:_encode_fields()
	local fl = ui._make_flags()

	ui._shift_flag(fl, self._disabled)
	ui._shift_flag(fl, self._on_press)

	return ui._encode("SZ", ui.Elem._encode_fields(self), ui._encode_flags(fl))
end

ui.Button._handlers[0x00] = function(self, ev, data)
	return self._on_press
end

ui.Toggle = ui._new_type(ui.Elem, "toggle", 0x03, true)

ui.Check  = ui.derive_elem(ui.Toggle, "check")
ui.Switch = ui.derive_elem(ui.Toggle, "switch")

function ui.Toggle:_init(props)
	ui.Elem._init(self, props)

	self._disabled = props.disabled
	self._selected = props.selected

	self._on_press = props.on_press
	self._on_change = props.on_change
end

function ui.Toggle:_encode_fields()
	local fl = ui._make_flags()

	ui._shift_flag(fl, self._disabled)
	ui._shift_flag_bool(fl, self._selected)

	ui._shift_flag(fl, self._on_press)
	ui._shift_flag(fl, self._on_change)

	return ui._encode("SZ", ui.Elem._encode_fields(self), ui._encode_flags(fl))
end

ui.Toggle._handlers[0x00] = function(self, ev, data)
	return self._on_press
end

ui.Toggle._handlers[0x01] = function(self, ev, data)
	local selected = ui._decode("B", data)
	ev.selected = selected ~= 0

	return self._on_change
end

ui.Option = ui._new_type(ui.Elem, "option", 0x04, true)

ui.Radio = ui.derive_elem(ui.Option, "radio")

function ui.Option:_init(props)
	ui.Elem._init(self, props)

	self._disabled = props.disabled
	self._family = props.family

	self._selected = props.selected

	self._on_press = props.on_press
	self._on_change = props.on_change

	if self._family then
		assert(ui.is_id(self._family), "Element family must be an ID string")
	end
end

function ui.Option:_encode_fields()
	local fl = ui._make_flags()

	ui._shift_flag(fl, self._disabled)
	ui._shift_flag_bool(fl, self._selected)

	if ui._shift_flag(fl, self._family) then
		ui._encode_flag(fl, "z", self._family)
	end

	ui._shift_flag(fl, self._on_press)
	ui._shift_flag(fl, self._on_change)

	return ui._encode("SZ", ui.Elem._encode_fields(self), ui._encode_flags(fl))
end

ui.Option._handlers[0x00] = function(self, ev, data)
	return self._on_press
end

ui.Option._handlers[0x01] = function(self, ev, data)
	local selected = ui._decode("B", data)
	ev.selected = selected ~= 0

	return self._on_change
end
